Minimum viable module source

guybedford opened the pull request:
Here's a first pass at what we've been discussing with respect to a minimum viable module source.

We define `%AbstractModuleSource%` as an intrinsic base class for module source class instances. It defines the `toStringTag` getter returning the internal `[[SourceModuleRecord]]`'s `[[SourceClassName]]` string (where `[[SourceModuleRecord]]` also has a `[[HostSourceMetadata]]` slot). This allows carrying host-context through this object back to the compilation phase with an abstract method that can be called by other specs within `HostLoadImportedModule`, where `[[HostDefined]]` on the instance record can be recreated on new instances in this way.

A new `GetModuleSource` object is used to provide the module source via Cyclic Module Record, and can be redefined by suitable cyclic module subclasses as necessary.

Feedback very welcome!

ljharb commented on README.md:
```diff
@@ -69,29 +69,41 @@ Only the `source` import source phase is specified by this proposal.
+    throw new Error();
+  }
+  get [Symbol.toStringTag] () {
+    if (this instanceof ModuleSource)
```
the semantics should match a private field, and testing whether that field is `in this`, rather than `instanceof`.

also, shouldn't this be defined in AbstractModuleSource, and the ModuleSource implementation more look like:
```js
#kind = 'ModuleSource'
```
thus deferring everything to the AbstractModuleSource getter?

guybedford replied:
I've changed the check in `Symbol.toStringTag` for `ModuleSource` to explicitly verify the host internal slot of `[[HostSourceData]]`.

ljharb commented on spec.emu:
```diff
@@ -844,3 +836,109 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+        <emu-alg>
+          1. Throw a *TypeError* exception.
+        </emu-alg>
+        <p>The *"length"* property of this function is *+0*<sub>𝔽</sub>.</p>
```
```suggestion
```
this is implied by the lack of any named arguments

ljharb commented on spec.emu:
```diff
@@ -844,3 +836,109 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+          <dd>Used to retrieve the host-defined data from a subclass of %AbstractModuleSource% to populate for a new instance of its module record by the host.</dd>
+        </dl>
+        <emu-alg>
+          1. Assert: _anyModuleSource_ is an instance of a subclass of %AbstractModuleSource%.
```
i'm not sure "instance of a subclass of" is a thing that's defined in the spec.

guybedford replied:
Thanks, removed this terminology.

ljharb commented on spec.emu:
```diff
@@ -844,3 +836,109 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+      <ul>
+        <li>is <dfn>%ModuleSource%</dfn>.</li>
+        <li>is the initial value of the *"ModuleSource"* property of the global object.</li>
+        <li>is a subclass of `%AbstractModuleSource%`.</li>
```
i think you want to use phrasing similar to https://tc39.es/ecma262/#sec-aggregate-error-constructor

guybedford replied:
Thanks, I've updated the wording.

ljharb commented on spec.emu:
```diff
@@ -844,3 +836,109 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+      <ul>
+        <li>is <dfn>%ModuleSource.prototype%</dfn>.</li>
+        <li>has a [[Prototype]] internal slot whose value is %AbstractModuleSource%.</li>
+        <li>has the following properties:</li>
```
```suggestion
        <li>is not a ModuleSource instance or an AbstractModuleSource instance and does not have a [[HostSourceMetadata]] internal slot.
        <li>has the following properties:</li>
```

guybedford commented:
@nicolo-ribaudo I've updated this PR to use your phase generalization model.

kriskowal commented on README.md:
```diff
@@ -69,29 +69,31 @@ Only the `source` import source phase is specified by this proposal.
-`WebAssembly.Module`, as defined in the [WebAssembly JS integration
-API][wasm-js-api].
+For WebAssembly modules, the existing `WebAssembly.Module` object is to be
+updated to be a subclass of `AbstractModuleSource` in the
```
I don’t see `AbstractModuleSource` here. Should there be a reference? I’m looking for the object graph you propose so one can walk by property from `globalThis` to `AbstractModuleSource`. I assume directly and that `ModuleSource` is a subclass, or prefer entirely for æsthetic reasons that I hold lightly, the omission of `AbstractModuleSource` in favor of `globalThis.ModuleSource` serving as both the base class for all language module sources and also the JavaScript module source. I will be satisfied for purposes of review by either detail.

guybedford replied:
I originally had an explicit `AbstractModuleSource`, but then saw we have existing precedent for prototype objects which are not themselves constructors (intrinsics ending in `Prototype`), so thought we could try that model instead.

So the chain is now effectively:

```js
global.ModuleSource = function ModuleSource () {
  throw new Error('unimplemented');
}
ModuleSource.prototype = ModuleSourcePrototype;
```

where ModuleSourcePrototype is then just defined as an object.

kriskowal replied:
That’s fine too, though prototypes usually participate in a `constructor.prototype` property cycle.

ljharb commented on spec.emu:
```diff
@@ -835,12 +1710,264 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+        <emu-alg>
+          1. Let _O_ be ? ToObject(*this* value).
+          1. Perform ? RequireInternalSlot(_R_, [[HostSourceData]]).
+          1. Return the String value *"ModuleSource"*.
```
Hardcoding the value in this method means that I can't use the base class method on every subclass to answer "which", only to answer "is". If, instead, we had a _second_ internal slot that contained "ModuleSource" by default, then subclasses would be encouraged to set the slot to their own value, and the base accessor would answer both "is" and "which" for all subclasses.

guybedford replied:
This method is unique to the JS source. Other sources would need to hard-code their own values, as is mentioned explicitly in the spec note.

Note that ModuleSource is not supposed to be subclassed by other sources, unless it were a true derivation of JavaScript source text somehow.

Creating a special slot to store a string is an option as well. Although we would still have to just imply conformance by telling other source types that they should change this slot, so it doesn't seem that much of a win as conformance would still be by convention?

ljharb replied:
Right - that's the point tho. I don't want to have to extract N accessors for N module types, I want one accessor to Rule Them All. In other words, this only really works if there exists a base class for all module types. If it's not ModuleSource, then we'd need AbstractModuleSource or similar.

kriskowal replied:
Is it satisfying if the base class has an accessor that responds based on an internal slot of the instance, such that the base class is not abstract but this behavior can be overridden? I don’t anticipate any JavaScript-specific properties on the module source base class, only properties with sensible semantics for any language of module.

guybedford replied:
You can still construct such a generic function by reading the combination of the getter functions across the list of supported subclasses right?

We can support a generic `toStringTag` on `ModuleSourcePrototype`, the main thing would be a new internal slot to store the exact source type string, which I'm just not sure is enough to justify the change.

ljharb replied:
The contents of the slot are expected to be overridden; ideally no module instance types would need to implement this accessor at all as a result - but if they did, then `baseAccessor.call(instance)` would continue to expose the slot value.

guybedford replied:
I feel like in this case, we should use something more like a `get type () -> String` over a `Symbol.toStringTag` though.

ljharb replied:
That's fine with me also.

kriskowal replied:
I like that `type` destructures and restructures with import attributes.
```js
const { type } = source;
await import(name, { with: { type } });
```
Or, at least it would if `type` were `undefined` for JavaScript, since `type` must be absent on the memo key for ECMAScript modules.

ljharb replied:
`{ with: { ...(type && { type }) } }`

kriskowal replied:
Apropos of this, @erights attributes the aphorism “Things that are similar should be the same or different”, to Alan Kay but it doesn’t appear to be well-known. Regardless, maybe `ModuleSource.prototype.language` would be better since it’s similar but not the same as the `type` cache key.

guybedford replied:
`sourceType`?

guybedford replied:
Ok, I've implemented `sourceType` using a `ModuleSourceRecord` abstraction to avoid two separate slots. And with JS modules of type `sourceModule.sourceType === "source-text"`.

nayeemrmn replied:
> would be better since it’s similar but not the same as the `type` component of a module memo key.

I wrote about this in https://github.com/tc39/proposal-import-attributes/issues/140, pointing out that the `type` import attribute is still mostly an assertion. Ideally this could be `type` and the attribute would be renamed to `assertType`.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -132,6 +920,9 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
             1. <ins>If _moduleSourceObject_ is *null*, then</ins>
               1. <ins>Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; a newly created *TypeError* object &raquo;).</ins>
               1. <ins>Return ~unused~.</ins>
+            1. <ins>Assert: _moduleSourceObject_.[[Prototype]] is %AbstractModuleSourcePrototype%.</ins>
```
I think the hierarchy here is wrong. It's asserting that
```js
import source s from "./mod.wasm";
Object.getPrototypeOf(s) === %AbstractModuleSourcePrototype%;
```

what I would expect is that every module source type has its own prototype class, and they all inherit from `%AbstractModuleSourcePrototype%`:

```js
import source s from "./mod.wasm";

// These equalities are already true with the current WebAssembly.Module class
s instanceof WebAssembly.Module;
Object.getPrototypeOf(s) === WebAssembly.Module.prototype;
s.constructor === WebAssembly.Module
Object.hasOwn(s, "constructor") === false;
Object.hasOwn(Object.getPrototypeOf(s), "constructor") === true;

// These equalities are new
Object.getPrototypeOf(Object.getPrototypeOf(s) === %AbstractModuleSourcePrototype%;
s instanceof Object.assign(function(){}, { prototype: %AbstractModuleSourcePrototype% })
```

i.e. it should be like this:
```js
WebAssembly.Module = function () { ... }
Object.setPrototypeOf(WebAssembly.Module, %AbstractModuleSourcePrototype%);
```
and not like this:
```js
WebAssembly.Module = function () { ... }
WebAssembly.Module.prototype = %AbstractModuleSourcePrototype%;
```

---

EDIT: What I'm suggesting is already what you wrote in the "The ModuleSource Constructor" section. It's just wrong in this assertion, where it should probably be

```suggestion
            1. <ins>Assert: _moduleSourceObject_.[[Prototype]].[[Prototype]] is %AbstractModuleSourcePrototype%.</ins>
```

guybedford replied:
Fixed, thanks.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -132,6 +920,9 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
               1. <ins>Return ~unused~.</ins>
+            1. <ins>Assert: _moduleSourceObject_.[[Prototype]] is %AbstractModuleSourcePrototype%.</ins>
+            1. <ins>Assert: _moduleSourceObject_ has a [[HostSourceData]] internal slot.</ins>
+            1. <ins>Assert: _moduleSourceObject_ is a ModuleSource instance if and only if _module_ is a Source Text Module Record.</ins>
```
This assertion is only valid if we enforce it somewhere (assertions are non-normative, they only assert existing internal properties of the spec). We could do it in the description of Module Record's [[ModuleSourceObject]] internal slot, or in the requirements of the HostLoadImportedModule hook (the former is better, because the hook is not called for the top-level module).

guybedford replied:
See updated assertions and description.

nicolo-ribaudo reviewed (commented):
I still have to catch up with the discussion that lead to the `.sourceType` accessor (https://github.com/tc39/proposal-import-reflection/pull/36#discussion_r1181971402), but in the meanwhile I want to write down what I believe was the conclusion from yesterday. I also left some other comments that are mostly unrelated to this.

---

Yesterday we agreed that:
- all module sources should extend form the same base class `AbstractModuleSource` (name tbd? I don't remember if it was the final name).
- `AbstractModuleSource` is "abstract", i.e. it cannot be directly instantiated
- `AbstractModuleSource.prototype[@@toStringTag]` is a getter that throws when it's called on a receiver that doesn't have the `[[ThisIsAModuleSource]]` internal slot (name tbd)
- `AbstractModuleSource.prototype[@@toStringTag]` would return the string `"[object AbstractModuleSource]"`
- All the "concrete" Module Source classes should define their own `[@@toStringTag]` prototype property, with `AbstractModuleSource.prototype[@@toStringTag]` returning `"[object AbstractModuleSource]"` being the fallback for when it's not shadowed (thanks to prototypical inheritance)

This satisfies the following goals:
- it is be possible to brand-check module sources (i.e. is this object a module source?) using the `AbstractModuleSource.prototype[@@toStringTag]` getter
- if a library cares about opaquely handling module sources, it is able to do so without having to use `XxxModuleSource`/`YyyModuleSource`-specific APIs
- if a library only wants to support `XxxModuleSource` but not `YyyModuleSource`, it doesn't need to hard-code any `YyyModuleSource`-specific methods. However, it should still use `XxxModuleSource`-specific methods (for example, to check that the module source it received is actually an `XxxModuleSource` and not something else, or for example to handle custom Wasm instantiation).

In practice, this is expressed by the following implementation:
```js
class AbstractModuleSource {
  #isModuleSource;

  constructor() {
    if (new.target === AbstractModuleSource) {
      throw new Error("Abstract class, you must extend it");
  }

  get [Symbol.toStringTag]() {
    if (!(#isModuleSource in this)) throw new Error("Invalid receiver");
    return "[object AbstractModuleSource]";
  }
}

WebAssembly.Module = class Module extends AbstractModuleSource {
  [Symbol.toStringTag] = "WebAssembly.Module";
}

class ModuleSource extends AbstractModuleSource {
  [Symbol.toStringTag] = "ModuleSource";
}
```

There was then the discussion about how to expose `AbstractModuleSource.prototype[@@toStringTag]` through ECMA-262 objects, with two different alternatives:
1. The proposal only defines `AbstractModuleSource`, and it's exposed as a property of the global object
2. The proposal defines both `AbstractModuleSource` and (for JS) `ModuleSource`, and only `ModuleSource` is exposed as a property of the global object. `AbstractModuleSource.prototype` would still be reachable using `Object.getPrototypeOf(ModuleSource.prototype)`, similarly to how it's for `AsyncFunction`&co.

---

We never discussed about only having `AbstractModuleSourcePrototype` and not the `AbstractModuleSource` function itself, but if there is already a precedent for this and if we go with option (2) it's a neat simplification that I might like.

As an aside, I really dislike the `source-text` string for JS modules because any language that is not compiled to a binary blob is "source text".

ljharb commented:
@nicolo-ribaudo not precisely; the best approach for my use cases is if the base abstract class has a getter that returns the string contents of a slot, and subclasses override the slot but *not* the getter.

guybedford commented on spec.emu:
```diff
@@ -132,6 +920,9 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
             1. <ins>If _moduleSourceObject_ is *null*, then</ins>
               1. <ins>Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; a newly created *TypeError* object &raquo;).</ins>
               1. <ins>Return ~unused~.</ins>
+            1. <ins>Assert: _moduleSourceObject_.[[Prototype]] is %AbstractModuleSourcePrototype%.</ins>
```
I've loosened the assertions, and also provided them in the description of the [[ModuleSourceObject]] slot.

ljharb commented on spec.emu:
```diff
@@ -835,12 +1757,262 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+      <h1>
+        CreateModuleSource (
+          _sourceClassName_: a String,
+          _hostDefined_: anything,
```
```suggestion
          _hostDefined_: unknown,
```

guybedford replied:
ECMA-262 does use the term `anything` for `_hostDefined_` on scripts and modules in the spec currently. This would need to be a whole spec edit to change I think.

ljharb replied:
ah ok, if it's already used then obv we'd stick with it here :-)

ljharb commented on spec.emu:
```diff
@@ -835,12 +1757,262 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+          1. Let _O_ be the *this* value.
+          1. If _O_ is not an Object, return *undefined*.
+          1. If _O_ does not have a [[ModuleSourceRecord]] internal slot, return *undefined*.
+          1. Let _name_ be _O_.[[ModuleSourceRecord]].[[SourseClassName]]
```
```suggestion
          1. Let _name_ be _O_.[[ModuleSourceRecord]].[[SourceClassName]]
```

nicolo-ribaudo commented on README.md:
```diff
@@ -69,29 +69,31 @@ Only the `source` import source phase is specified by this proposal.
-which moves the construction of module records to the host, and on which we can define a
-`[[ModuleSourceObject]]` custom object to be returned by module source phase that is compatible
-with module expressions.
+The object provided by the module source phase must have a prototype set to the
+`ModuleSourcePrototype` object, defined by this specification to be a minimal
+shared base prototype for a compiled modular resource.
+
+In addition we define the `sourceType` getter returning the type string
+corresponding to the name of the specific module source subclass.
```
```suggestion
The object provided by the module source phase must have a the
`AbstractModuleSourcePrototype` object in its prototype chain, defined by this
specification to be a minimal shared base prototype for a compiled modular resource.
```

lucacasonato replied:
Also include something along the lines of:

For now, the `%AbtractModuleSourcePrototype%` contains only a `@toStringTag` getter.

nicolo-ribaudo commented on README.md:
```diff
@@ -69,29 +69,31 @@ Only the `source` import source phase is specified by this proposal.
 
-In the current specification, the JS module source case just throws an error
-to support this as a future addition.
+It's `sourceType` value would be `"source-text"`.
 
```
```suggestion
```

nicolo-ribaudo commented on README.md:
```diff
@@ -69,29 +69,31 @@ Only the `source` import source phase is specified by this proposal.
+Future proposals may then add support for [bindings lookup methods][],
+the [ModuleSource constructor][] and [instantiation][] support.
+
+New properties may be added to the base `ModuleSourcePrototype`, or shared
+with ECMASCript module sources via `ModuleSource` additions.
```
```suggestion
New properties may be added to the base `AbstractModuleSourcePrototype`, or shared
with ECMAScript module sources via `ModuleSource.prototype` additions.
```

nicolo-ribaudo commented on README.md:
```diff
@@ -69,29 +69,31 @@ Only the `source` import source phase is specified by this proposal.
-still being able to support the same CSP policy as the native [ESM
-integration][wasm-esm], avoiding the need for `unsafe-wasm-eval` for custom Wasm
-execution.
+For WebAssembly modules, the existing `WebAssembly.Module` object is to be
+updated to have a prototype value of `ModuleSourcePrototype` in the
```
```suggestion
For WebAssembly modules, the existing `WebAssembly.Module.prototype` object is to be
updated to have a prototype value of `AbstractModuleSourcePrototype` in the
```

nicolo-ribaudo commented on spec.emu:
```diff
@@ -290,15 +1188,8 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
-                  <p>For <a href="https://webassembly.github.io/esm-integration/js-api/index.html#esm-integration">WebAssembly modules</a>, this field should contain the corresponding `WebAssembly.Module` object.</p>
-                </emu-note>
+                <ins>The ModuleSource Object corresponding to this source Module Record's source phase (<emu-xref href="#sec-module-source-objects"></emu-xref>), if it has been requested.</ins>
+                <ins>This object must have a [[ModuleSourceRecord]] internal slot and %AbstractModuleSourcePrototype% at the root of its prototype chain.</ins>
```
```suggestion
                <ins>This object must have a [[ModuleSourceRecord]] internal slot and %AbstractModuleSourcePrototype% in its prototype chain.</ins>
```

`%AbstractModuleSourcePrototype%` has `Object.prototype` as it's [[Prototype]], so it's not the root :P

lucacasonato commented:
Just want to make sure we are all aware that there will be two `@toStringTag` for `WebAssembly.Module`. One will be on `%AbstractModuleSource%` (the one defined here). The other will be on `WebAssembly.Module.prototype` (defined by WebIDL). The latter will shadow the former. Both will return `WebAssembly.Module`. One can always access the former one through `WebAssembly.Module.prototype.prototype[Symbol.toStringTag]`.

lucacasonato commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# Module Source Imports
+# Import Source Reflection
```
Let's rename to `Source Phase Imports`?

guybedford replied:
I've come around on the `Reflection` term finally - in fact the `ModuleSource` option is defined in the `Reflection` section now. The `ModuleSource` once it has bindings will truly be a reflective data type. `Import Source Reflection` allows us to somewhat salvage the original name.

lucacasonato replied:
I think it's confusing still. I'd prefer "phase".

guybedford replied:
Sure, `Source Phase Imports` makes sense, and but strictly speaking, we are treating `ModuleSource` as a reflection object now.

kriskowal replied:
I prefer to reserve `Reflection` for the proposal that makes module sources reflect the parsed bindings. I propose that `Source Phase Imports` is clearer for this layer.

lucacasonato commented on README.md:
````diff
@@ -50,7 +50,7 @@ const x = await import("<specifier>", { phase: "source" });
 ```
 
 For dynamic imports, import phase is specified in the same second
-attribute options bag that [import assertions][] are specified in, using the
+attribute options bag that [import attributes][] are specified in, using the
 `phase` key.
````
It is not immediately clear from this sentence that `phase` is not _within_ the attributes bag. Maybe rephrase or add an example?

guybedford replied:
There is an example before, but I've updated the wording to be clearer.

lucacasonato commented on README.md:
```diff
@@ -69,29 +69,31 @@ Only the `source` import source phase is specified by this proposal.
-which moves the construction of module records to the host, and on which we can define a
-`[[ModuleSourceObject]]` custom object to be returned by module source phase that is compatible
-with module expressions.
+The object provided by the module source phase must have a prototype set to the
+`ModuleSourcePrototype` object, defined by this specification to be a minimal
+shared base prototype for a compiled modular resource.
```
```suggestion
The object provided by the module source phase should have a prototype set to the
`%AbtractModuleSourcePrototype%` object, defined by this specification to be a minimal
shared base prototype for a compiled modular resource.
```

lucacasonato commented on README.md:
```diff
@@ -69,29 +69,31 @@ Only the `source` import source phase is specified by this proposal.
+Future proposals may then add support for [bindings lookup methods][],
+the [ModuleSource constructor][] and [instantiation][] support.
+
+New properties may be added to the base `ModuleSourcePrototype`, or shared
```
```suggestion
New properties may be added to the base `%AbstractModuleSourcePrototype%`, or shared
```

nicolo-ribaudo reviewed (commented):
There is probably some unclosed `<emu-clause>` tag; the nesting is off:
<img width="472" alt="image" src="https://user-images.githubusercontent.com/7000710/235901886-9d0c19f9-7855-4494-a20b-135ef5386ee4.png">

(more comments coming)

nicolo-ribaudo commented on spec.emu:
```diff
@@ -502,54 +1376,148 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+              1. Set _index_ to _index_ + 1.
+              1. Append _module_ to _stack_.
+              1. For each <del>String</del><ins>ModuleRequest</ins> _required_ of _module_.[[RequestedModules]], do
+                1. <ins>If _required_.[[Phase]] is ~evaluation~, then</ins>
```
We should fail linking if the source is not available, right?

```
1. Else,
  1. Assert: _required_.[[Phase]] is ~source~.
  1. If _requiredModule_.[[ModuleSourceObject]] is *null*, throw a SyntaxError.
```

(SyntaxError feels wrong, but it's what we use for other linking errors)

guybedford replied:
Thanks, I've made this a `ReferenceError` for now.

lucacasonato commented on README.md:
```diff
@@ -119,29 +121,14 @@ In turn this enables [Wasm components to be able to import][]
 
-Other module types may define their own host module sources. If no module phase import
-is defined, it will fail during the loading phase.
+Any other host-defined module types may define their own host module sources. If no module phase import is defined, it will fail during the loading phase.
```
This sentence doesn't seem clear. Maybe:

```suggestion
Any other host-defined module types may define their own host module sources. If a given module does not define a source representation for it's source, importing it with a "source" phase target fails.
```

guybedford replied:
Sounds good, thanks!

nicolo-ribaudo reviewed (commented):
ℹ️ These comments are more part of my stage 3 review than to the minimum viable module source, feel free to tackle them in a separate PR if you prefer.

---

The spec is now missing a fundamental piece: we are not actually exposing the source imported by `import source x from "./mod"` anywhere!

In `InitializeEnvironment`, we should add inside the loop
```
1. If _in_.[[ModuleRequest]].[[Phase]] is ~source~, then
  1. Let _source_ be _importedModule_.[[ModuleSourceObject]].
  1. Assert: _source_ is not *null*, otherwise it would have failed while linking.
  1. Perform ! _env_.CreateImmutableBinding(_in_.[[LocalName]], true).
  1. Perform ! _env_.InitializeBinding(in.[[LocalName]], _source_).
```

Also, maybe it would make sense to store ~source~ instead of "default" in an ImportEntry Record's [[ImportName]] when importing the source.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -669,27 +1596,22 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
                 1. Let _namespace_ be GetModuleNamespace(_importedModule_).
                 1. Perform ! _env_.CreateImmutableBinding(_in_.[[LocalName]], *true*).
                 1. Perform ! _env_.InitializeBinding(_in_.[[LocalName]], _namespace_).
-              1. <ins>Else if _in_.[[ImportName]] is ~module~, then</ins>
-                1. <ins>Let _moduleSourceObject_ be _importedModule_.[[ModuleSourceObject]].</ins>
-                1. <ins>If _moduleSourceObject_ is *null*, throw a *TypeError* exception.</ins>
-                1. <ins>Perform ! _env_.CreateImmutableBinding(_in_.[[LocalName]], *true*).</ins>
-                1. <ins>Perform ! _env_.InitializeBinding(_in_.[[LocalName]], _moduleSourceObject_).</ins>
```
I just realized that my review above (https://github.com/tc39/proposal-import-reflection/pull/36#pullrequestreview-1410694415) is exactly what has been deleted here. What is the reason for deleting it?

kriskowal commented on README.md:
```diff
@@ -49,9 +49,7 @@ not supported.
-For dynamic imports, import phase is specified in the same second
-attribute options bag that [import assertions][] are specified in, using the
-`phase` key.
+For dynamic imports, import phase is specified as a separate `phase` key in the options object to dyamic import, similar to how [import attributes][] use the `attributes` key.
```
Not the `"with"` key?

guybedford replied:
Thanks will fix.

kriskowal commented on README.md:
```diff
@@ -69,29 +67,29 @@ Only the `source` import source phase is specified by this proposal.
-`[[ModuleSourceObject]]` custom object to be returned by module source phase that is compatible
-with module expressions.
+The object provided by the module source phase should have a prototype set to the
+`%AbtractModuleSourcePrototype%` object, defined by this specification to be a minimal
```
`Abstract` may be an unnecessary qualifier in this name, since this is to be or become the value of `ModuleSource.prototype` (or `ModuleSource.prototype.__proto__`? Is the intention to allow `ModuleSource.prototype` to be the seat of JavaScript specific features and `ModuleSource.prototype.__proto__ === %AbstractModuleSourcePrototype%`?).

I know we can brand-check module sources by `Object.getOwnPropertyDescriptor(ModuleSource.prototype, Symbol.toStringTag).get.call(subject) !== undefined`, but can one check whether something is an `instanceof` an abstract module source? I think that’s not possible without an abstract module source constructor, like `subject instanceof AbstractModuleSource`, or rather `subject instanceof ModuleSource.prototype.__proto__.constructor`.

guybedford replied:
@kriskowal yes and yes. So `ModuleSource.prototype.__proto__ === %AbstractModuleSourcePrototype%` and `Object.getOwnPropertyDescriptor(ModuleSource.prototype.__proto__, Symbol.toStringTag).get.call(subject) !== undefined` and there's no `AbstractModuleSource` constructor to check.

I'm open to making `AbstractModuleSource` more explicit with a constructor.

kriskowal replied:
I don’t think we need a `globalThis.AbstractModuleSource === %AbstractModuleSourceConstructor%` and I think it’s sufficient for:
- `globalThis.ModuleSource.prototype.__proto__ === %AbstractModuleSourcePrototype%`,
- `%AbstractModuleSourcePrototype%.constructor === %AbstractModuleSourceConstructor%`, and
- `%AbstractModuleSourceConstructor%.prototype === %AbstractModuleSourcePrototype%`,
such that `subject instanceof ModuleSource.prototype.__proto__.constructor` can discriminate module source instances with only slightly less fidelity than a brand check.

guybedford replied:
Sure, I can add an AbstractModuleSource constructor for this kind of check at least.

guybedford replied:
I've added an AbstractModuleSource constructor in the latest commit.

kriskowal commented on README.md:
```diff
@@ -69,29 +67,29 @@ Only the `source` import source phase is specified by this proposal.
-which moves the construction of module records to the host, and on which we can define a
-`[[ModuleSourceObject]]` custom object to be returned by module source phase that is compatible
-with module expressions.
+The object provided by the module source phase should have a prototype set to the
+`%AbtractModuleSourcePrototype%` object, defined by this specification to be a minimal
```
```suggestion
The object provided by the module source phase must be an object that inherits the
`%AbtractModuleSourcePrototype%` object prototypically, defined by this specification to be a minimal
```

kriskowal replied:
Or “must be an object with `%AbstractModuleSourcePrototype%` in its chain of prototypes”.

kriskowal commented on README.md:
```diff
@@ -69,29 +67,29 @@ Only the `source` import source phase is specified by this proposal.
+`%AbtractModuleSourcePrototype%` object, defined by this specification to be a minimal
+shared base prototype for a compiled modular resource.
+
+In addition we define the `sourceType` getter returning the type string
+corresponding to the name of the specific module source subclass.
```
I believe we’re converging on omitting this in favor of an `Symbol.toStringTag` getter on the `%AbstractModuleSourcePrototype%` that is sufficient to distinguish the module source brand and also differentiate implementations by consulting internal slots. A `toStringTag` of `"ModuleSource"` would imply source text for this language, as opposed to `"WebAssembly.Module"`.

For purposes of consensus, I am not picky about these particulars. None of the motivating use cases I am tracking require a brand check or module source type reflection. Module source objects can be opaque tokens for import virtualization and only need to reveal their language-agnostic `bindings` for loader virtualization, bundling, or import graph analysis.

guybedford replied:
Thanks, this was also missed in the refactor.

kriskowal commented on README.md:
```diff
@@ -69,29 +67,29 @@ Only the `source` import source phase is specified by this proposal.
-The type of the module source object for a JavaScript module is would depend
-on the [module source][] specification.
+For JavaScript modules, the module source phase is then specified to return
+a `ModuleSource` object, representing an ECMAScript Module Source.
```
Is this the intention?
```suggestion
a `ModuleSource` object, representing an ECMAScript Module Source.
The `ModuleSource.prototype` is a `%ModuleSourcePrototype%`.
The `[[Proto]]` of `%ModuleSourcePrototype%` is `%AbstractModuleSourcePrototype%`.
```

kriskowal commented on README.md:
```diff
@@ -69,29 +67,29 @@ Only the `source` import source phase is specified by this proposal.
-still being able to support the same CSP policy as the native [ESM
-integration][wasm-esm], avoiding the need for `unsafe-wasm-eval` for custom Wasm
-execution.
+For WebAssembly modules, the existing `WebAssembly.Module.prototype` object is to be
+updated to have a prototype value of `AbstractModuleSourcePrototype` in the
```
Is this the intent?
```suggestion
For WebAssembly modules, the existing `WebAssembly.Module.prototype` object is to be
updated to have a `[[Proto]]` of `AbstractModuleSourcePrototype` in the
```

kriskowal commented on README.md:
```diff
@@ -119,29 +117,14 @@ In turn this enables [Wasm components to be able to import][]
 
-Other module types may define their own host module sources. If no module phase import
-is defined, it will fail during the loading phase.
+Any other host-defined module types may define their own host module sources. If a given module does not define a source representation for it's source, importing it with a "source" phase target fails.
```
Suggest:

Host-defined module sources must include `%AbstractModuleSourcePrototype%` in their prototype chain and support the *necessary internal slots for revealing their `toString` tag to a brand check, the underlying module source record, and host data*

guybedford replied:
Thanks, added along these lines.

ljharb commented on README.md:
```diff
@@ -77,6 +77,8 @@ corresponding to the name of the specific module source subclass.
 
 For JavaScript modules, the module source phase is then specified to return
 a `ModuleSource` object, representing an ECMAScript Module Source.
+The `ModuleSource.prototype` is a `%ModuleSourcePrototype%`.
```
nit: p sure neither of these should have articles
```suggestion
`ModuleSource.prototype` is `%ModuleSourcePrototype%`.
```

guybedford replied:
I think I already covered this.

nicolo-ribaudo commented:
I have some other observations about what ECMA-262 currently does.

1. (@kriskowal) I don't particularly care about whether or not we have the `%AbstractModuleSource%` constructor, other than for a general tendency of reducing useless objects. The precedent goes in both directions:
   - `%ArrayIteratorPrototype%`, `%AsyncFromSyncIteratorPrototype%`, `%AsyncIteratorPrototype%`, `%IteratorPrototype%`, `%MapIteratorPrototype%`, `%RegExpStringIteratorPrototype%`, `%SetIteratorPrototype%`, `%StringIteratorPrototype%` are all constructor-less prototypes.
   - `%AsyncFunction%`, `%AsyncGeneratorFunction%` and `%GeneratorFunction%` are all non-exposed constructors (there isn't just the prototype), however they can actually be used to eval code.
   - `%TypedArray%` is the only non-exposed non-constructable constructor that exists, where just having the prototype object would have been enough.
   - the [Itertor Helpers](https://tc39.es/proposal-iterator-helpers/) proposal introduces a new `%Iterator%` constructor (exposed as `globalThis.Iterator`), so that user-defined classes can `extends Iterator`.

   I have a slight preference here for not having `%AbstractModuleSource%` yet, and only add it if we'll need it for user-defined subclasses (i.e. by the virtual module sources proposal), but also just having it here following the TypedArray precedent is fine.
2. (@lucacasonato) the Itertor Helpers introduces a new [`Iterator.prototype[@@toStringTag]`](https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype-@@tostringtag) property, and all the built-in iterator subclasses overwrite it. So that will create another place where we have multiple `@@toStringTag`s in the prototype chain of an object, although in that case they would have different values.

nicolo-ribaudo commented on README.md:
```diff
@@ -174,13 +161,16 @@ needed. See the security improvements section for more details.
     https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
 [ECMA-262 ES modules HostLoadImportedModule refactoring]:
     https://github.com/tc39/ecma262/pull/2905
+[ModuleSource constructor]: 
```
Is this link missing the URL?

nicolo-ribaudo commented on spec.emu:
```diff
@@ -4,11 +4,799 @@
+              <td>
+                <ins>%ModuleSource%</ins>
+              </td>
+              <td>
+              </td>
```
```suggestion
              <td>
                `ModuleSource`
              </td>
```

I thought we were going to expose `ModuleSource` on the global object.

nicolo-ribaudo replied:
It's _not_ listed here as exposed, we should list it if it is (in the currently empty column).

nicolo-ribaudo commented on spec.emu:
```diff
@@ -187,33 +1022,48 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+          1. <del>For each String _name_ of _additionalNames_, do</del>
+          1. <ins>For each ModuleRequest Record _mr_ of _additionalRequests_, do</ins>
+            1. <ins>Let _found_ be *false*.</ins>
+            1. <ins>For each ModuleRequest Record _mr2_ of _requests_, do</ins>
+              1. <ins>If _mr_.[[Specifer]] is _mr2_.[[Specifer]] and _mr_.[[Phase]] is _mr2_.[[Phase]], then</ins>
+                1. <ins>Assert: _found_ is *false*.</ins>
+                1. <ins>Set _found_ to *true*.</ins>
+            1. <del>If _moduleNames_ does not contain _name_, then</del>
+            1. <ins>If _found_ is *false*, then</ins>
+              1. Append <del>_name_</del><ins>_mr_</ins> to <del>_moduleNames_</del><ins>_requests_</ins>.
```
Since `ModuleRequests` is used to populate the `[[RequestedModules]]` field of module records, and since we want to load each module once even if it has different phases, we should deduplicate this list. If two modules have the same specifier but different phases, we should still set _found_ to true (so that don't we append it again) and just upgrade the phase in the existing ModuleRequest Record to max(phase1, phase2).

I believe that the reason you removed the de-duplication logic is because we still want the duplicates in ImportEntries, because if we have `import source s from "./a.js"; import d from "./a.js"` we want to bind both `s` and `d`. It's true that ImportEntries internally calls ModuleRequests, however:
- it only calls ModuleRequests on pieces of individual import statements, and thus it wouldn't go through this deduplication logic. It's similar to how in
  ```js
  import a from "./a.js";
  import b from "./a.js";
  ```
  we have two different ImportEntries but only one ModuleRequest.
- if necessary to make it more clear, you could make it not use ModuleRequests anymore and instead do something like https://nicolo-ribaudo.github.io/proposal-defer-import-eval/#sec-static-semantics-importentries

nicolo-ribaudo commented on spec.emu:
```diff
@@ -290,15 +1187,8 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
-                  <p>For <a href="https://webassembly.github.io/esm-integration/js-api/index.html#esm-integration">WebAssembly modules</a>, this field should contain the corresponding `WebAssembly.Module` object.</p>
-                </emu-note>
+                <ins>The ModuleSource Object corresponding to this source Module Record's source phase (<emu-xref href="#sec-module-source-objects"></emu-xref>), if it has been requested.</ins>
+                <ins>This object must have a [[ModuleSourceRecord]] internal slot and %AbstractModuleSource.prototype% in its prototype chain.</ins>
```
```suggestion
                <ins>This object must have a [[ModuleSourceRecord]] internal slot and should have %AbstractModuleSource.prototype% in its prototype chain.</ins>
```

per https://github.com/tc39/proposal-import-reflection/pull/36#discussion_r1182533387

nicolo-ribaudo commented on spec.emu:
```diff
@@ -502,54 +1375,158 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
             <emu-alg>
               1. Assert: _state_.[[IsLoading]] is *true*.
-              1. <ins>Let _importOnlyAsPhase_ be *false*.</ins>
-              1. <ins>If _referrer_ has been specified, then</ins>
-                1. <ins>Assert: _specifier_ has been specified.</ins>
```
We should still use `_referrer_` and `_specifier_` to know if we are in a "source phase import" and thus stop recursively loading the dependencies. As the spec is written in this PR, we load all the dependencies of source imports.

Something like this:
```
1. Let _loadDependencies_ be *true*.
1. If _referrer_ has been specified, then:
  1. Assert: _specifier_ has been specified.
  1. Let _request_ be the only ModuleRequest Record in _referrer_.[[RequestedModules]] such that _request_.[[Specifier]] is _specifier_.
  1. If _request_.[[Phase]] is ~source~, set _loadDependencies_ to *false*.
```

We still need the changes to `ContinueModuleLoading` to propagate referrer and specifier.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -502,54 +1375,158 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+        </emu-clause>
+
+        <emu-clause id="sec-moduledeclarationlinking" type="concrete method" oldids="sec-moduledeclarationinstantiation">
+          <h1>Link ( ): either a normal completion containing ~unused~ or a throw completion</h1>
```
You need to add the `<dl>` header to get proper rendering of the return type. Also in `Evaluate`.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -502,54 +1375,158 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+              1. Append _module_ to _stack_.
+              1. For each <del>String</del><ins>ModuleRequest</ins> _required_ of _module_.[[RequestedModules]], do
+                1. <ins>If _required_.[[Phase]] is ~source~, then</ins>
+                  1. <ins>If _requiredModule_.[[ModuleSource]] is *null*, throw a *ReferenceError* exception.</ins>
```
```suggestion
                  1. <ins>If _requiredModule_.[[ModuleSourceObject]] is *null*, throw a *ReferenceError* exception.</ins>
```

nicolo-ribaudo replied:
Also, now that we validate this while linking you can replace step 7.c.ii of InitializeEnvironment with an assertion that the source is not null.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -835,12 +1711,179 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+        <li>throws when called as a constructor.</li>
+        <li>also throws when called as a function rather than as a constructor.</li>
+        <li>has a *"length"* property whose value is *0*<sub>𝔽</sub>.</li>
+        <li>has a [[Prototype]] internal slot whose value is %Function.prototype%.</li>
```
The definition of `[[Prototype]]` is duplicated here and in _"Properties of the %AbstractModuleSource% Intrinsic Object"_.

Given that this object cannot be really used as a constructor, I would just copy what `%TypedArray%` does (https://tc39.es/ecma262/#sec-%typedarray%-intrinsic-object).

nicolo-ribaudo commented on spec.emu:
```diff
@@ -835,12 +1711,179 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+      <ul>
+        <li>is <dfn>%ModuleSource%</dfn>.</li>
+        <li>is the initial value of the *"ModuleSource"* property of the global object.</li>
+        <li>has a [[Prototype]] internal slot whose value is %AbstractModuleSource.prototype%.</li>
```
```suggestion
        <li>has a [[Prototype]] internal slot whose value is %AbstractModuleSource%.</li>
```

```js
class A {}
class B extends A {}
B.__proto__ === A;
```

nicolo-ribaudo replied:
I realize now that this is a good reason to have `%AbstractModuleSource%`. All the case I listed in a comment above where the constructor doesn't exist are never extended by an existing constructor.

guybedford replied:
//cc @kriskowal do you have any strong opinions here with regards to retaining an `AbstractModuleSource` constructor explicitly existing.

kriskowal replied:
I don’t have any strong opinions about these details.

guybedford replied:
Okay, so this is a strong argument for the %AbstractModuleSource% constructor as the shared home of static methods for module sources per standard classes in JS. Agreed this model works out at this point, so that seems to settle the `AbstractModuleSource` structuring question.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -835,12 +1711,179 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+
+      <emu-clause id="sec-modulesource.prototype.constructor">
+        <h1>ModuleSource.prototype.constructor</h1>
+        <p>The initial value of `ModuleSource.prototype.constructor` is ModuleSource.</p>
```
```suggestion
        <p>The initial value of `ModuleSource.prototype.constructor` is %ModuleSource%.</p>
```

nicolo-ribaudo reviewed (commented):
In [ParseModule](https://tc39.es/ecma262/#sec-parsemodule) we should create the `[[ModuleSourceObject]]` for JS modules.

guybedford commented:
@nicolo-ribaudo I've handled the requests problem by piping the `ModuleRequest` record through the host resolve hooks.

Instead of explicitly creating `ModuleSourceObject`, the expectation of the `HostLoadImportedModule` hook is that when the phase is `source`, it must then populate the `ModuleSourceObject`. If it doesn't we then throw the reference error.

As far as I can tell this refactoring does in fact better align with import attributes. It also seems useful to provide the phase information to the host hook, so far as the preloader needs to be aware of the phasing, which seems critical information as well.

Some other minor edits were needed to get the piping to work, there may well be further future simplifications on this stuff (eg `ModuleSource Records` as first class host hooks, resources / internal registry as first class keys), but for now this seems like a reasonable compromise on the design while aligning with the other spec changes for the most part.

Let me know your thoughts further!

guybedford commented:
Per @lucacasonato's feedback I've removed the global `ModuleSource` for now in this PR. His argument was that defining an object that does not have any use at all would not be reasonable at this point. At the very least we can still define `AbstractModuleSource` and make this reachable through `WebAssembly.Module` for the WebAssembly use case. JS sources would then continue to throw for now during link time, while having a path for support in future.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -502,28 +690,28 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+                1. <del>Set _state_.[[PendingModulesCount]] to _state_.[[PendingModulesCount]] + _requestedModulesCount_.</del>
+                1. <del>For each String _required_ of _module_.[[RequestedModules]], do</del>
+                1. <ins>For each ModuleRequest Record _required_ of _module_.[[RequestedModules]], do</ins>
+                  1. If _module_.[[LoadedModules]] contains a Record whose [[Specifier]] is _required_<ins>.[[Specifier]]</ins>, then
```
This PR now duplicates entries in the `[[ReqeustedModules]]` list in case there are different phases (i.e. `import source s from "x"; import x from "x"`), however this check still skips calling `HostLoadImporteedModule` is that module has already been loaded even if for a different phase. Is this intentional?

nicolo-ribaudo reviewed (commented):
I talked about this pull request with @guybedford privately, specifically about the signature of `HostLoadImportedModule`, and there is a disagreement between us that we have not yet been able to solve.

_(tl;dr: scroll down to the end of this comment, after the `<hr>`, for concrete next steps)_

I do not think that this proposal should change the signature of that host hook, and that its parameters should continue being:
- `referrer`: the Module Record of the module that contains the import
- `specifier`: a String representing the import (e.g. `"foo"` in `import { x } from "foo"`). The import attributes proposal will extend this to be a `(specifier, attributes)` pair, because attributes are necessary to identify the requested module as much as the specifier string.
- `hostDefined`: some host-specific data, opaque to ECMA-262, to be propagated across the whole loading process. This is currently needed by HTML (step 3 of https://html.spec.whatwg.org/#fetch-the-descendants-of-and-link-a-module-script), but I would eventually like to try refactoring the HTML spec to not rely on this.
- `payload`: some ECMA-262-specific data, opaque to the host, that is needed by ECMA-262 to continue the loading process after that the host finishes loading the requested module.

Module requests (a fancy word used by the spec to describe the intention behind import statements) with the same specifier should be deduplicated, and we should call the host hook once per `(specifier, referrer)` pair.

@guybedford would prefer to change the hook signature as follows:
- `referrer`: the Module Record of the module that contains the import
- `(specifier, phase)`: a pair containing the string specifier of the import and the corresponding phase (with this proposal, `~source~` or `~evaluation~`).
- `hostDefined`: some host-specific data, opaque to ECMA-262, to be propagated across the whole loading process. This is currently needed by HTML (step 3 of https://html.spec.whatwg.org/#fetch-the-descendants-of-and-link-a-module-script), but I would eventually like to try refactoring the HTML spec to not rely on this.
- `payload`: some ECMA-262-specific data, opaque to the host, that is needed by ECMA-262 to continue the loading process after that the host finishes loading the requested module.

Module requests with the same specifier _and with the same phase_ should be deduplicated, and we should call the host hook once per `(specifier, referrer, phase)` triple.

My motivation is that, as we have discussed/discovered when deeply analyzing what "phases" are and how they are different from import attributes, the phase should not affect how an individual module is loaded and thus we should not pass it to the host. This is exactly the same reason as to why we will pass attributes but not the phase to the import hook in the [compartments](https://github.com/tc39/proposal-compartments/blob/master/0-module-and-module-source.md) proposal. The responsibility of handling the import phase is completely on ECMA-262, and we should avoid blurring the layering line.

Guy motivations are that:
1. Passing the phase would allow the host to not populate the `[[ModuleSourceObject]]` slot of the returned module record unless it's going to be used, so that we avoid allocating an unnecessary object.
2. Browsers have a preloader that eagerly loads dependencies of modules, to optimize loading times. This preloader needs to know if a module is "fully imported" or if we are only importing its source, to avoid creating HTTP requests for its dependencies unless they are necessary.
3. The phase is needed by ECMA-262 to resume the loading process, so we need to pass it through.

I believe (1) has been solved by replacing `[[ModuleSourceObject]]` with a `GetModuleSourceObject()` "getter". I also believe that this was not necessary, because allocating an object is unobservable and thus implementations could have already deferred that work until the `[[ModuleSourceObject]]` internal slot was accessed, implementing it internally as a getter. However, I don't dislike the current solution since most of the other extension points of Module Record are already based on abstract methods overwritten by the subclasses.

For (3), we already have the `payload` whose purpose is exactly that.

My answer for (2) is probably the one that has been less convincing, but:
- this preloader is not defined anywhere; neither in ECMA-262 nor in HTML. It's something defined internally in the engines, that is not observable (unless you observe the HTTP requests received by your server, obviously). Host hooks are meant to provide an interface for host specs, or to provide an integration API for ECMA-262 (I know that some delegates disagree in this second point, don't hold me accountable on this :P). Implementations that have magic optimizations are still free to read any internal state defined in the ECMA-262 spec, even if that's not officially exposed through host hooks, unless these optimizations would have observable effects from within ECMA-262.
- the preloader is already speculative and based on heuristics, and it not always "correct" (where by correct I mean "do the optimal amount of work; no more and no less"). As an example, if a long file has a bunch of import statements at the top and a syntax error at the bottom, the preloader might start loading the dependencies declared in those import statements even if ECMA-262 will never call HostLoadImportedModule to load them. There would need to be one more heuristic to skip preloading dependencies of modules imported just as module sources (potentially, since the dependencies might still be needed in case later the source is instantiated with the `Module` constructor from compartments), but this heuristic can be based on data not directly exposed by ECMA-262 to host specs.

Additionally, given that module types that we initially expect to work with this proposal (i.e. `WebAssembly.Module`, and nothing else) do not support loading dependencies thorugh the current module loader, there will be no preloaded involved with their dependencies. If passing the information about the phase ends up being necessary (and I don't think it will, due to the reasons above) we can always update the host hook signature once there is a concrete need for it.

---

I recommended to @guybedford to merge this PR even if I disagree with this layering interface, because this proposal is going to be presented at the next TC39 meeting in one week and we should have the observable semantics merged as soon as possible so that other delegates can read them in advance.

However, I re-reviewed the changes pushed in the past few commits and I don't think that this PR can be merged as-is (read my comments below as for why).

@guybedford would you be ok with reverting all the commits after https://github.com/tc39/proposal-import-reflection/pull/36/commits/fab8784b45c056e9d0bee37eb0a515bdf66752c1 (i.e. `git reset --hard fab8784b`), only delete the `ModuleSource` constructor on top of that, merge this PR, and re-open a separate PR with the remaining changes where we can continue the discussion?

littledan commented:
How are we doing on making the call of `sourceType` vs `@@toStringTag` type checks? I'd be OK with either outcome personally (as long as we don't have multiple strong type checks).

guybedford commented:
@littledan the way that ended up last week was that we now have a `toStringTag` getter function on `AbstractModuleRecord` which checks a `[[SourceClassName]]` internal slot from the object and returns that as its value, exactly mirroring TypedArray as suggested by @nicolo-ribaudo.

ljharb commented:
(most builtins have multiple slot checks, but any additional ones added to module things would, i hope, have an actual purpose beyond that)

guybedford commented on spec.emu:
```diff
@@ -502,28 +690,28 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+                    1. <ins>If _required_.[[Phase]] is not ~source~, then</ins>
+                      1. Perform InnerModuleLoading(_state_, _record_.[[Module]]).
                   1. Else,
+                    1. <ins>Set _state_.[[PendingRequestsCount]] to _state_.[[PendingRequestsCount]] + 1.</ins>
```
Thanks for bringing up the sync use case here, I wasn't aware of this. I've gone back to the original algorithm on the iteration.

guybedford commented on spec.emu:
```diff
@@ -545,119 +732,189 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
-                1. Append _ee_ to _indirectExportEntries_.
-            1. Let _async_ be _body_ Contains `await`.
-            1. Return Source Text Module Record { [[Realm]]: _realm_, [[Environment]]: ~empty~, [[Namespace]]: ~empty~, [[CycleRoot]]: ~empty~, [[HasTLA]]: _async_, [[AsyncEvaluation]]: *false*, [[TopLevelCapability]]: ~empty~, [[AsyncParentModules]]: &laquo; &raquo;, [[PendingAsyncDependencies]]: ~empty~, [[Status]]: ~new~, [[EvaluationError]]: ~empty~, [[HostDefined]]: _hostDefined_, [[ECMAScri ...
+            1. Let _moduleSourceObject_ be _module_.[[ModuleSourceObject]].
```
I've gone ahead with your suggestion of not defining `[[ModuleSourceObject]]` slot for now, and instead working entirely through `GetModuleSource()`. I've then included a note on this function that it should maintain idempotency.

guybedford commented on spec.emu:
```diff
@@ -545,119 +732,189 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+            1. If _moduleSourceObject_ is *null*, then
+              1. 1. Return ThrowCompletion(a newly created *ReferenceError* object).
+            1. If _moduleSourceObject_ is not an Object, does not have a [[ModuleSourceRecord]] internal slot, or if _moduleSourceObject_.[[Prototype]].[[Prototype]] is not %AbstractModuleSource.prototype%, then
+              1. 1. Return ThrowCompletion(a newly created *TypeError* object).
```
I've removed this error, and instead used a note.

guybedford commented:
@nicolo-ribaudo I believe I have addressed all of your feedback, I think it would be useful to carefully distinguish editorial suggestions from normative changes at this point in the review process. Specifically editorial changes can be discussed after this PR is merged, but I know you'll agree getting this merged is the priority at this point.

I will go ahead and land this today, unless there are any further objections.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -63,76 +63,37 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
-          1. Let _state_ be the DynamicImportState Record { [[PromiseCapability]]: _promiseCapability_, [[SourcePhase]]: _sourcePhase_ }.
-          1. Perform HostLoadImportedModule(_referrer_, _specifierString_, ~empty~, _state_).
+              1. Set _phase_ to ~source~.
+              1. Perform HostLoadImportedModule(_referrer_, _specifierString_, ~empty~, _promiseCapability_).
```
```suggestion
```

nicolo-ribaudo reviewed (approved):
_Reluctantly_ approving 😛 

After this PR is merged, I plan to propose some changes to the layering. I still have to figure exactly what would satisfy me while still meeting your goals.

I just left a comment deleting a line that you probably copy-pasted from an older version of the PR, if @lucacasonato could commit it and merge this PR it would help us moving quicker.

nicolo-ribaudo commented:
@erights @kriskowal I added exposing `%AbstractModuleSource%` to the agenda for the next SES meeting, if that's ok.

lucacasonato commented on spec.emu:
```diff
@@ -732,12 +890,62 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+          </li>
+          <li>
+            <p><ins>When `[[Phase]]` is set to ~source~, this should not affect the identity or state of the resultant Module Record as if it had been called for the ~evaluation~ phase.</ins></p>
+            <p><ins>In addition, the ~source~ phase indicates that preloading should not take place in the host for this module.</ins></p>
```
"Preloading" is not an ECMA262 thing, so this should be a note (ie not normative).

I still question the need for passing phase but this does not impact observable semantics to users, as long as there is note that identity and load is not affected. Let's consider this in a follow up.

lucacasonato reviewed (approved):
Landing because it's semantically correct now.

@nicolo-ribaudo will address some editorial things in a follow up.

bakkot commented:
Heads up that 262 editors haven't had / probably aren't going to have time to review this before next week, so it's probably not going to be able to achieve stage 3 at this meeting for that reason.
